### PR TITLE
Create ETL that joins crashes and Moped components

### PR DESCRIPTION
**Pull request by Charlie-Henry**
## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/19453

## Testing

**Steps to test:**
First fill out the env_template for Hasura, from 1pass. Make sure you use Moped **production** and VZ **staging** credentials 

Note, if you are on Apple Silicon you may need to add `--platform linux/amd64` to get GDAL to install correctly.

```
docker build . -t atddocker/vz-moped-join
```
```
docker run -it --env-file env_file atddocker/vz-moped-join /bin/bash
python moped_project_components_spatial_join.py
```

The logging should be pretty self explanatory to see if things succeeded or not.

You can also then check if `moped_component_crashes` was populated with the lookup table. (maybe delete everything in that table yourself to double check if the script did anything).

---

#### Ship list

- [ ] Check migrations for any conflicts with latest migrations in `main` branch
- [ ] Confirm Hasura role permissions for necessary access
- [ ] Code reviewed
- [ ] Product manager approved

### CODE REVIEW

File: etl/moped_projects/moped_project_components_spatial_join.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    )["crashes"]
+    logger.info(f"{len(crashes)} crash records retrieved")
+
+    components = load_json_to_df(components, geom_col="geometry")
+    crashes = load_json_to_df(crashes, geom_col="position")
+
+    # Transforming to projected coordinates
+    # NAD83 / Texas Central (ftUS)
+    components = components.to_crs(epsg=2277)
+    crashes = crashes.to_crs(epsg=2277)
```
Comment (johnclary) [inline etl/moped_projects/moped_project_components_spatial_join.py:82]: i love how easy this code is to follow!

**Review by johnclary — APPROVED** [review]
Awesome—this worked like a charm for me 🙌 

I have a suggestion about the crash query filter and the python version we're using. Otherwise this looks fantastic!!!

Thanks, Charlie! 🚢 🚢 🚢 🚢

File: etl/moped_projects/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.10-slim
```
[the remark this answers is not in the document]
Comment (Charlie-Henry) [reply etl/moped_projects/Dockerfile]: 3.12 worked ✅

File: etl/moped_projects/queries.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Gets the ID and location of every record in the crashes table
+query_vz = """
+    {
+        crashes(where: { position: { _is_null: false } }) {
```
[the remark this answers is not in the document]
Comment (Charlie-Henry) [reply etl/moped_projects/queries.py]: I've added your changes to the vz query 👍

**Review by chiaberry — APPROVED** [review]
`2024-11-13 16:36:06,309 INFO: 511591 rows uploaded...`

The logging is nice, being able to follow whats happening. 🚢

File: etl/moped_projects/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+mobility projects and crashes.
+
+`moped_project_components_spatial_join.py` buffers the project geometry (points and lines) to a distance of 40 feet
+then spatially joins the crashes that are inside of that polygon.
+
+## Docker
+
+It is recommended to run this script using the docker container. You can build it using:
+
+Note, if you are on Apple Silicon you may need to add `--platform linux/amd64` to get GDAL to install correctly.
```
Comment (mddilley) [inline etl/moped_projects/README.md:14]: 🙏 thanks for this detail because this was the case for me

File: etl/moped_projects/queries.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        position
+      }
+    }
+"""
+
+# Gets the ID and geometry for every record in the component_arcgis_online_view table
+query_moped = """
+    {
+        component_arcgis_online_view(where: { geometry: { _is_null: false } }) {
+            project_component_id,
```
Comment (mddilley) [inline etl/moped_projects/queries.py:21]: Commas not creating a syntax error in GraphQL is a TIL for me. Cool to know! https://spec.graphql.org/June2018/#sec-Insignificant-Commas

File: etl/moped_projects/moped_project_components_spatial_join.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    # Spatial join
+    logger.info(f"Buffering project geometries to a distance of {BUFFER_DISTANCE} feet")
+    components["geometry"] = components.geometry.buffer(BUFFER_DISTANCE)
+    components.set_index("project_component_id", inplace=True)
+    crashes.set_index("id", inplace=True)
+    logger.info(f"Joining crashes spatially to project geometry")
+    crashes_near_projects = gpd.sjoin(crashes, components, how="inner")
```
Comment (mddilley) [inline etl/moped_projects/moped_project_components_spatial_join.py:90]: nice! this is a pretty powerful line, and a [TIL](https://geopandas.org/en/stable/gallery/spatial_joins.html#Inner-join) for me. 😎

**Review by mddilley — APPROVED** [review]
I tested locally, and I saw the staging table truncate and fill back up with records. Always fun to learn more about geo python! 🙌 🚀

**Pull request merged**